runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java

```diff
@@ -789,13 +844,19 @@ public Instant get() {
    */
   public static <ExecutableT, CollectionT>
       WatermarkManager<ExecutableT, ? super CollectionT> create(
-          Clock clock, ExecutableGraph<ExecutableT, ? super CollectionT> graph) {
-    return new WatermarkManager<>(clock, graph);
+          Clock clock,
+          ExecutableGraph<ExecutableT, ? super CollectionT> graph,
+          Function<ExecutableT, String> getName) {
```

Comment:
Optional: Do you think we should just rely on toString on ExecutableT to provide watermark metadata information?

nit: rename `getName` to `toName`

Reply:
Unfortunately, that would not work for the use case in this PR, which uses PTransformNode::getId. That provides a more concise, comprehensible string for debugging and is more appropriate for this use case.

Reply:
Yup, what Ben said. toString() for the types used here tends to be multiple kilobytes long and contains protos.